File: backend/app/schemas/employee.py
"""
StriveNest ERP - Employee Management Pydantic Schemas
Data transfer objects and validation models for Employee CRUD, Profile, Documents, Approval, and Audit History.
"""

from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, EmailStr, Field, ConfigDict


# --- Employee Schemas ---

class EmployeeBase(BaseModel):
    first_name: str = Field(..., min_length=1, max_length=100)
    last_name: str = Field(..., min_length=1, max_length=100)
    email: EmailStr
    phone: Optional[str] = Field(None, max_length=20)
    department: Optional[str] = Field(None, max_length=100)
    department_id: Optional[str] = Field(None, max_length=36)
    designation: Optional[str] = Field(None, max_length=100)
    designation_id: Optional[str] = Field(None, max_length=36)
    role: str = Field("EMPLOYEE", max_length=50)
    profile_image: Optional[str] = Field(None, max_length=500)
    status: str = Field("PENDING", max_length=50)
    is_verified: bool = False


class EmployeeCreate(EmployeeBase):
    password: Optional[str] = Field(None, min_length=6, max_length=128)
    employee_id: Optional[str] = Field(None, max_length=50)


class EmployeeUpdate(BaseModel):
    first_name: Optional[str] = Field(None, min_length=1, max_length=100)
    last_name: Optional[str] = Field(None, min_length=1, max_length=100)
    email: Optional[EmailStr] = None
    phone: Optional[str] = Field(None, max_length=20)
    department: Optional[str] = Field(None, max_length=100)
    department_id: Optional[str] = Field(None, max_length=36)
    designation: Optional[str] = Field(None, max_length=100)
    designation_id: Optional[str] = Field(None, max_length=36)
    role: Optional[str] = Field(None, max_length=50)
    profile_image: Optional[str] = Field(None, max_length=500)
    status: Optional[str] = Field(None, max_length=50)
    is_active: Optional[bool] = None
    is_verified: Optional[bool] = None


class EmployeeSelfUpdate(BaseModel):
    first_name: Optional[str] = Field(None, min_length=1, max_length=100)
    last_name: Optional[str] = Field(None, min_length=1, max_length=100)
    phone: Optional[str] = Field(None, max_length=20)
    profile_image: Optional[str] = Field(None, max_length=500)


class EmployeeStatusUpdate(BaseModel):
    status: str = Field(..., max_length=50, description="Status: PENDING, ACTIVE, INACTIVE, BLOCKED, REJECTED, DELETED")
    reason: Optional[str] = Field(None, max_length=500)


class EmployeeResponse(BaseModel):
    id: str
    employee_id: str
    first_name: str
    last_name: str
    full_name: str
    email: str
    phone: Optional[str] = None
    role: str
    department_id: Optional[str] = None
    department: Optional[str] = None
    designation_id: Optional[str] = None
    designation: Optional[str] = None
    profile_image: Optional[str] = None
    status: str
    is_active: bool
    is_verified: bool
    created_at: str
    updated_at: str
    last_login: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class EmployeePaginatedResponse(BaseModel):
    total: int
    page: int
    size: int
    pages: int
    items: List[EmployeeResponse]


class EmployeeStatsSummaryResponse(BaseModel):
    total_employees: int
    active_employees: int
    pending_employees: int
    inactive_employees: int
    blocked_employees: int
    deleted_employees: int
    verified_employees: int
    recent_employees: List[EmployeeResponse]


# --- Document Schemas ---

class DocumentCreate(BaseModel):
    document_type: str = Field(..., min_length=1, max_length=100)
    document_name: str = Field(..., min_length=1, max_length=255)
    file_name: str = Field(..., min_length=1, max_length=255)
    file_url: str = Field(..., min_length=1, max_length=500)
    file_size: int = Field(..., ge=0)
    mime_type: str = Field(..., min_length=1, max_length=100)


class DocumentResponse(BaseModel):
    id: str
    employee_id: str
    document_type: str
    document_name: str
    file_name: str
    file_url: str
    file_size: int
    mime_type: str
    status: str
    uploaded_by: Optional[str] = None
    created_at: str
    updated_at: str

    model_config = ConfigDict(from_attributes=True)


class DocumentPaginatedResponse(BaseModel):
    total: int
    page: int
    size: int
    items: List[DocumentResponse]


# --- Audit History Schemas ---

class EmployeeAuditResponse(BaseModel):
    id: str
    employee_id: str
    performed_by: Optional[str] = None
    performer_name: Optional[str] = None
    action: str
    details: Optional[str] = None
    ip_address: Optional[str] = None
    created_at: str

    model_config = ConfigDict(from_attributes=True)


class EmployeeAuditPaginatedResponse(BaseModel):
    total: int
    page: int
    size: int
    items: List[EmployeeAuditResponse]

File: superadmin/src/types/employee.ts
export type EmployeeStatus = 'PENDING' | 'ACTIVE' | 'INACTIVE' | 'BLOCKED' | 'REJECTED' | 'DELETED';
export type UserRole = 'SUPER_ADMIN' | 'EMPLOYEE';

export interface Employee {
  id: string;
  employee_id: string;
  first_name: string;
  last_name: string;
  full_name: string;
  email: string;
  phone?: string | null;
  role: string;
  department_id?: string | null;
  department?: string | null;
  designation_id?: string | null;
  designation?: string | null;
  profile_image?: string | null;
  status: EmployeeStatus;
  is_active: boolean;
  is_verified: boolean;
  created_at: string;
  updated_at: string;
  last_login?: string | null;
}

export interface EmployeeCreatePayload {
  first_name: string;
  last_name: string;
  email: string;
  phone?: string;
  department?: string;
  designation?: string;
  role?: string;
  status?: string;
  password?: string;
  employee_id?: string;
  is_verified?: boolean;
}

export interface EmployeeUpdatePayload {
  first_name?: string;
  last_name?: string;
  email?: string;
  phone?: string;
  department?: string;
  designation?: string;
  role?: string;
  status?: string;
  is_active?: boolean;
  is_verified?: boolean;
  profile_image?: string;
}

export interface EmployeeFilterParams {
  page?: number;
  size?: number;
  search?: string;
  department?: string;
  designation?: string;
  role?: string;
  status?: string;
  is_active?: boolean;
  is_verified?: boolean;
  include_deleted?: boolean;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface EmployeeDocument {
  id: string;
  employee_id: string;
  document_type: string;
  document_name: string;
  file_name: string;
  file_url: string;
  file_size: number;
  mime_type: string;
  status: string;
  uploaded_by?: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentCreatePayload {
  document_type: string;
  document_name: string;
  file_name: string;
  file_url: string;
  file_size: number;
  mime_type: string;
}

export interface EmployeeAuditLog {
  id: string;
  employee_id: string;
  performed_by?: string | null;
  performer_name?: string | null;
  action: string;
  details?: string | null;
  ip_address?: string | null;
  created_at: string;
}

export interface PaginatedResponse<T> {
  total: number;
  page: number;
  size: number;
  pages?: number;
  items: T[];
}

export interface EmployeeStatsSummary {
  total_employees: number;
  active_employees: number;
  pending_employees: number;
  inactive_employees: number;
  blocked_employees: number;
  deleted_employees: number;
  verified_employees: number;
  recent_employees: Employee[];
}
